Corona-Warn-App/src/main/java/de/rki/coronawarnapp/dccreissuance/ui/consent/DccReissuanceConsentViewModel.kt

```diff
@@ -11,46 +11,59 @@ import de.rki.coronawarnapp.covidcertificate.common.certificate.DccV1
 import de.rki.coronawarnapp.covidcertificate.person.core.PersonCertificatesProvider
 import de.rki.coronawarnapp.covidcertificate.person.core.PersonCertificatesSettings
+import de.rki.coronawarnapp.dccreissuance.core.processor.DccReissuanceProcessor
 import de.rki.coronawarnapp.tag
 import de.rki.coronawarnapp.util.coroutine.DispatcherProvider
 import de.rki.coronawarnapp.util.ui.SingleLiveEvent
 import de.rki.coronawarnapp.util.viewmodel.CWAViewModel
 import de.rki.coronawarnapp.util.viewmodel.CWAViewModelFactory
 import kotlinx.coroutines.flow.catch
+import kotlinx.coroutines.flow.first
 import kotlinx.coroutines.flow.map
 import timber.log.Timber
 
 class DccReissuanceConsentViewModel @AssistedInject constructor(
     dispatcherProvider: DispatcherProvider,
     personCertificatesProvider: PersonCertificatesProvider,
-    private val dccQrCodeExtractor: DccQrCodeExtractor,
-    private val format: CclTextFormatter,
     @Assisted private val personIdentifierCode: String,
+    private val format: CclTextFormatter,
+    private val dccQrCodeExtractor: DccQrCodeExtractor,
+    private val reissuanceProcessor: DccReissuanceProcessor,
     private val personCertificatesSettings: PersonCertificatesSettings,
+    private val dccSwapper: DccSwapper,
 ) : CWAViewModel(dispatcherProvider) {
 
     internal val event = SingleLiveEvent<Event>()
 
-    internal val stateLiveData: LiveData<State> =
-        personCertificatesProvider.personCertificates.map { certificates ->
-            val person = certificates.find { it.personIdentifier?.codeSHA256 == personIdentifierCode }
-            person?.personIdentifier?.let {
-                personCertificatesSettings.dismissReissuanceBadge(it)
-            }
+    private val reissuanceData = personCertificatesProvider.findPersonByIdentifierCode(personIdentifierCode)
+        .map { person ->
+            person?.personIdentifier?.let { personCertificatesSettings.dismissReissuanceBadge(it) }
             person?.dccWalletInfo?.certificateReissuance
-        }.map { certificateReissuance ->
-            certificateReissuance.toState()
-        }.catch {
+        }
+    internal val stateLiveData: LiveData<State> = reissuanceData.map { it.toState() }
+        .catch {
             Timber.tag(TAG).d(it, "dccReissuanceData failed")
+            event.postValue(Back) // Should not happen ¯\_(ツ)_/¯
         }.asLiveData2()
 
-    internal fun startReissuance() {
-        event.postValue(ReissuanceInProgress)
-        // call api
-        // replace certificate
-        // navigate
+    internal fun startReissuance() = launch {
+        runCatching {
+            event.postValue(ReissuanceInProgress)
+            val dccReissuanceDescriptor = reissuanceData.first()!!
+            reissuanceData.first()?.certificateToReissue
+            val response = reissuanceProcessor.requestDccReissuance(dccReissuanceDescriptor)
+            dccSwapper.swap(response, dccReissuanceDescriptor.certificateToReissue)
```

Comment:
What about doing the swapping in the `processor`? I don't like that we are having so much business logic directly in the `ViewModel`. Should help us with testing too, I think.

Reply:
I mean having a dependency to `DccSwapper` in the `processor` and calling `.swap()` there ...

Reply:
It was that way when Bernd implemented it , I think the processor should handle server call only , adding new certificate could also throw an error which is not related to server API

Reply:
Ok, I will create another class then, to make it simpler in here

Reply:
Adjusted